feat: Gator Permissions Revocation

hanzel98 opened the pull request:
## Overview

This PR adds a complete revocation flow for gator permissions, including delegation status checking and transaction handling.

## Key Changes

### 1. Enhanced Gator Permissions Controller
- **Added delegation status checking**: New `isDelegationDisabled` function to check if a delegation is already disabled on-chain
- **Improved error handling**: Better error messages and logging for debugging
- **Enhanced API**: Updated `fetchAndUpdateGatorPermissions` to accept optional parameters for filtering

### 2. Background API Integration
- **Added `call` method**: New API method in `metamask-controller.js` to handle `eth_call` requests
- **Network client integration**: Proper routing of RPC calls through the network controller
- **Messenger configuration**: Added `TransactionController:transactionConfirmed` event to allowed events

### 3. Delegation Utilities
- **New delegation functions**: Added `encodeDisabledDelegationsCheck` and `decodeDisabledDelegationsResult` for on-chain delegation status checking
- **Enhanced delegation handling**: Better support for checking delegation states

### 4. Revocation Flow Improvements
- **Smart revocation logic**: Automatically detects if delegation is already disabled and handles accordingly
- **Transaction management**: Proper handling of disable transactions and pending revocations
- **Better user experience**: More informative logging and error handling

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Enabled revocation flow of advance permissions

## Technical Details

### Files Modified:
- `app/scripts/metamask-controller.js`: Added `call` method for RPC requests
- `app/scripts/controller-init/messengers/gator-permissions/gator-permissions-controller-messenger.ts`: Added transaction confirmed event
- `ui/store/controller-actions/gator-permissions-controller.ts`: Enhanced with delegation checking and improved API
- `ui/hooks/gator-permissions/useRevokeGatorPermissions.ts`: Added delegation status checking
- `shared/lib/delegation/delegation.ts`: New delegation utility functions

### Key Features:
1. **On-chain delegation checking**: Verifies if delegation is already disabled before creating transactions
2. **Improved error handling**: Better debugging and user feedback
3. **Enhanced API flexibility**: Support for parameterized permission fetching
4. **Transaction flow optimization**: Smarter handling of revocation scenarios

## Testing
- All existing functionality preserved
- New delegation checking logic tested
- Error handling improved with better logging


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds full Gator permissions revocation flow with on-chain disabled checks, background/controller APIs, UI hooks, and config/policy updates.
> 
> - **Gator Permissions Controller/API**:
>   - Expose `api` methods: `addPendingRevocation`, `submitRevocation`, and parameterized `fetchAndUpdateGatorPermissions`.
>   - Add background handler `checkDelegationDisabled` via `metamask-controller.js` (performs `eth_call`), and wire to NetworkController.
>   - Messenger now delegates TransactionController events: `transactionConfirmed`, `transactionFailed`, `transactionDropped`.
> - **UI**:
>   - `useRevokeGatorPermissions`: resolve network client per permission, pre-check on-chain disabled state, call `addPendingRevocation`/`submitRevocation`, handle batch revocations (filter already-disabled), and navigate on pending confirmation.
>   - `useGatorPermissions`: call `fetchAndUpdateGatorPermissions({ isRevoked: false })`.
>   - Header navbar: handle intermediate Gator Permissions page before Sites.
>   - Controller actions: add `addPendingRevocation`, `submitRevocation`, `checkDelegationDisabled`; update `fetchAndUpdateGatorPermissions` to accept params.
> - **Shared utils**:
>   - Add `encodeDisabledDelegationsCheck` and `decodeDisabledDelegationsResult` for `disabledDelegations(bytes32)` calls.
> - **Config/Build**:
>   - Enable `GATOR_PERMISSIONS_REVOCATION_ENABLED` for Flask build.
>   - LavaMoat policy: allow `setTimeout`/`clearTimeout` for `@metamask/gator-permissions-controller`.
>   - Bump `@metamask/gator-permissions-controller` to `^0.4.0` (lockfile updated); update `@metamask/shield-controller`.
> - **Tests/State**:
>   - Update tests and snapshots; add `pendingRevocations` to state shape; e2e navigation adjusted.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 233483c4c55afa3e36b753c09ccc58b65dc19dc1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

hanzel98 committed:
chore: added already disabled delegation check, and filtered permissioned

hanzel98 committed:
chore: adding new allowed events

hanzel98 commented:
I have read the CLA Document and I hereby sign the CLA

hanzel98 committed:
chore: deleted console logs, added mocks

hanzel98 commented:
Right now the actions is failing because the gator-permissions-controller needs to be updated
<img width="1339" height="135" alt="image" src="https://github.com/user-attachments/assets/0130d7db-4eef-40d8-ad7e-61381fac078a" />

There are 2 dependencies
- https://github.com/MetaMask/core/pull/6713
- https://github.com/MetaMask/snap-7715-permissions/pull/171

MoMannn commented on ui/store/controller-actions/gator-permissions-controller.ts:
```diff
@@ -1,10 +1,69 @@
-    );
-  };
+export const fetchAndUpdateGatorPermissions = async (
+  params?: Json,
```
Maybe define params type? Probably define in core and import here so its not just Json?

hanzel98 commented on app/scripts/metamask-controller.js:
```diff
@@ -2201,6 +2201,19 @@ export default class MetamaskController extends EventEmitter {
     return {
       // etc
       getState: this.getState.bind(this),
+      call: async (method, params, networkClientId) => {
```
I made a different function but similar concept please let me know if it makes sense

No `this` context: these are standalone exported functions, not class methods.
Network controller doesn't exist in UI: this.networkController only exists in the background script.

hanzel98 commented on ui/hooks/gator-permissions/useRevokeGatorPermissions.ts:
```diff
@@ -213,6 +219,43 @@ export function useRevokeGatorPermissions({
+      if (isDisabled) {
+        await submitRevocation(permissionContext);
+        // Return a mock transaction meta since no actual transaction is needed
+        const mockTransactionMeta = {
```
Sure

hanzel98 commented on app/scripts/controller-init/gator-permissions/gator-permissions-controller-init.ts:
```diff
@@ -59,6 +59,8 @@ export const GatorPermissionsControllerInit: ControllerInitFunction<
     api: {
       fetchAndUpdateGatorPermissions:
         controller.fetchAndUpdateGatorPermissions.bind(controller),
+      addPendingRevocation: controller.addPendingRevocation.bind(controller),
```
Sure

hanzel98 committed:
chore: callRpc, deleted mock tx

jeffsmale90 committed:
Update controllers:
- @metamask/gator-permissios-controller
- @metamask/signature-controller
- @metamask/permissions-kernel-controller
- @metamask/shielf-controller

hanzel98 committed:
Merge branch 'main' into feat/add-revocation-flow

MoMannn commented:
@metamaskbot update-policies

MoMannn committed:
Fix import order.

MoMannn committed:
enable gator revocation in flask

MoMannn committed:
Fix tests

metamaskbot committed:
Update LavaMoat policies

MoMannn committed:
Test fixes

MoMannn committed:
refactord checking disabled delegations (#37540)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37540?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Replace generic RPC with dedicated `checkDelegationDisabled` flow,
wiring backend API and refactoring hook/tests to skip tx when delegation
already disabled.
> 
> - **Gator Permissions (backend)**:
> - Add `checkDelegationDisabled` API on `MetamaskController` to query
`disabledDelegations(bytes32)` via `eth_call` and decode result.
> - Expose `checkDelegationDisabled` in controller API; remove generic
`callRpc` exposure.
> - **UI Hook**:
> - `useRevokeGatorPermissions` now calls `checkDelegationDisabled`
before submitting a revoke tx; if already disabled, calls
`submitRevocation` and skips tx.
> - Update imports and mocks from `isDelegationDisabled` →
`checkDelegationDisabled`; adjust tests accordingly.
> - **Controller Actions**:
> - Add `checkDelegationDisabled` action delegating to background;
remove manual ABI encode/decode and generic RPC usage.
> - Introduce `FetchAndUpdateGatorPermissionsParams` with optional
`isRevoked`.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
0b78869ecc6274864f4568e6e2e3766cc3f271df. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

MoMannn committed:
update state snapshot

MoMannn committed:
update test to correctly navigate when gator permissions revocation flow in enabled

MoMannn committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-extension into feat/add-revocation-flow

MoMannn committed:
restore main package versions

MoMannn committed:
remove state changes in ApprovalController

hanzel98 committed:
fix: object types for gator controller functions

hanzel98 committed:
Merge branch 'main' into feat/add-revocation-flow

hanzel98 committed:
chore: network by permission chain id

hanzel98 committed:
fix: prevent navigation with undefined transactionId in useRevokeGatorPermissions

When revoking an already-disabled permission, the hook returns null and setTransactionId(null?.id) sets transactionId to undefined. The useEffect hook now checks that transactionId is defined before attempting navigation to avoid calling navigateToId(undefined).

hanzel98 committed:
Merge branch 'main' into feat/add-revocation-flow

MoMannn committed:
update unit tests

MoMannn committed:
Merge branch 'feat/add-revocation-flow' of https://github.com/MetaMask/metamask-extension into feat/add-revocation-flow

MoMannn committed:
remove unnecessary test

jeffsmale90 reviewed (approved):
Looks great!

davidmurdoch reviewed (approved):
Policy file changes look good.

As does the `metamask-controller.js` integration.

I didn't review beyond those domains.

jeffsmale90 committed:
Merge branch 'main' into feat/add-revocation-flow

seaona reviewed (approved):
QA changes LGMT
